### PR TITLE
[Gekidou MM-46229] Add hideAndLock, showAndUnlock callbacks for Search

### PR DESCRIPTION
**Pull request by jfrerich**
#### Summary
This PR fixes several issues with the navigation header. 

1) Hides and locks the header when the results screen is shown. This prevents a user from pulling down and exposing the title.
1) Adds padding margin to the results header so the `RoundedHeaderContext` rounded edges show.
1) allows locking the header to a lockValue via the `hideAndLock` callback
   1) sets `lockValue` to the padding value after hiding the header
1) allows unlocking the header via the `showAndUnlock` callback
   1) resets `lockValue` to `null` and unlocks the header
1) When showing results and `file filter` or `team` is changed, show the `Loading` component causing the results page to flicker. Just reload the results with the new data.  This keeps the results header constant and only updates the value for the number of files found and the filter badge.
   1) **(Not in figma)** could add a loading for the Results component which adds a loader within that window.  
   1) the initial loading component would show a loader including the header and results (already how this works)

- [x] verify show more working before performing search
- [x] verify show more working after a search and returns to initial screen (was causing a hang in the app)

#### Refactor
- remove insets.top from `scrollPaddingTop` hook value 
   - in components that used `scrollPaddingTop`, insets.top was being subtracted
- add `insets.top` to the `useDefaultHeader` hook result
   - in components that use this hook, they were adding the insets.top value 
   - now they do not depend on insets.top and get the exact value from the hook result
- components affected: 
   - `channel/header/header.tsx`
   - `channel/channel.tsx`
   - `gallery/header/index.tsx`
   - `global_threads/index.tsx`
   - `recent_mentions.tsx`
   - `saved_messages.tsx`
   - `post_draft/post_draft.tsx`

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-46229

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: 
* iPhone 14 (ios 16.4)
* iPad Pro 11-inch (ios 16.0)
* Pixel 5 API 30 (Android 11.0 Google APIs | x86) 

#### Release Note

```release-note

```

### CODE REVIEW

File: app/components/navigation_header/index.tsx [hunk 1]
```diff
@@ -58,58 +61,73 @@ const NavigationHeader = ({
...
+                        subtitle={subtitle}
+                        theme={theme}
+                        title={title}
+                        translateY={translateY}
+                    />
                 }
                 {hasSearch &&
                     <NavigationSearch
                         {...searchProps}
-                        defaultHeight={defaultHeight}
```
Comment (jfrerich) [inline app/components/navigation_header/index.tsx:107]: I added a new prop `topMargin` and calculate the value based on the props that are available at this level (`defaultHeight`, `largeHeight`, `scrollValue`)

File: app/components/navigation_header/index.tsx [hunk 2]
```diff
@@ -58,58 +61,73 @@ const NavigationHeader = ({
...
             height: Math.max(height, minHeight),
             minHeight,
             maxHeight: largeHeight + insets.top + MAX_OVERSCROLL,
         };
     });
 
+    const minScrollValue = useDerivedValue(() => scrollValue?.value || 0, [scrollValue]);
+    const heightOffset = useDerivedValue(() => largeHeight - defaultHeight, [largeHeight, defaultHeight]);
+
+    const translateY = useDerivedValue(() => (
```
Comment (jfrerich) [inline app/components/navigation_header/index.tsx:80]: I extracted these two calculations from their components and pass them in as props. I think the is easier to see how the new lockValue value is used for each compononent.

**Comment by jfrerich** [comment]
I have tested this on ipad and iphone, but android is failing to build on my end.  @matthewbirtch, this should resolve the flicking and get the clear and cancel options to operate as expected, with respect to expanding and collapsing the header.

**Comment by matthewbirtch** [comment]
@jfrerich is this ready for me to review now then?

**Comment by jfrerich** [comment]
> @jfrerich is this ready for me to review now then?

Yes.  I'd love to get your feedback.

**Comment by matthewbirtch** [comment]
@jfrerich this is working so much better now. A few things I noticed below, but seems a lot better.

1. The app crashes (on iOS simulator) when I'm in the collapsed state of the search top bar (focused on the search input) and then if I tap the  'Show more' button under 'Search options'. This is repeateable every time I follow this sequence.

https://user-images.githubusercontent.com/2040554/195441669-fd6b3229-d132-480b-baf0-5173e21e59d2.mov

2.  The autocomplete shows when I select one of my recent searches. This is likely something for another PR though - not necessarily caused by this one.

https://user-images.githubusercontent.com/2040554/195441890-22c5e7a3-d0d6-4579-8bb2-2354ec78ff51.mov

3. The margin on the empty state is too tight to the edges. Again, this could be fixed in a separate PR
image

**Comment by matthewbirtch** [comment]
> can you verify that you are using the latest commit in the PR? I was also facing the crashing issue, but I have it working on my end now.

@jfrerich I just pulled the latest and the app still crashes for me when I follow the steps I outlined

File: app/components/navigation_header/index.tsx [hunk 3]
```diff
@@ -58,58 +61,71 @@ const NavigationHeader = ({
...
+    ), [lockValue?.value, minScrollValue.value, headerOffset]);
+
+    const searchTopMargin = useAnimatedStyle(() => {
+        const margin = Math.min(-Math.min(minScrollValue.value, headerOffset), headerOffset);
+        const unlockedBottomMargin = margin + ViewConstants.UNLOCKED_SEARCH_BOTTOM_MARGIN;
+        const bottomMargin = lockValue?.value ? -lockValue?.value : unlockedBottomMargin;
+        return {
+            marginTop: bottomMargin - ViewConstants.SEARCH_INPUT_HEIGHT,
+        };
+    }, [lockValue?.value, headerOffset, minScrollValue.value]);
```
Comment (larkox) [inline app/components/navigation_header/index.tsx:90]: lockValue and minScrollValue don't need to be in the dependencies.

**Review by larkox — APPROVED** [review]
LGTM. No major problems, just a few nits.

It would be great to have a quick chat to see the motivation for some of these changes, just to understand a bit better what is happening in some parts :smile:

**Comment by matthewbirtch** [comment]
@jfrerich looks like the crash is now fixed, thanks!

Should we fix these other few issues in separate PRs or will you include them in this one?

> 2. The autocomplete shows when I select one of my recent searches. This is likely something for another PR though - not necessarily caused by this one.

> 3. The margin on the empty state is too tight to the edges. Again, this could be fixed in a separate PR

**Comment by jfrerich** [comment]
@matthewbirtch, I'm fine with merging, but need @enahum review first.

**Review by matthewbirtch — APPROVED** [review]
Approving this. We'll tackle the remaining identified issues in separate PRs

File: app/components/navigation_header/search.tsx [hunk 4]
```diff
@@ -2,35 +2,27 @@
...
 import {Events} from '@constants';
-import {HEADER_SEARCH_HEIGHT} from '@constants/view';
 import {changeOpacity, makeStyleSheetFromTheme} from '@utils/theme';
 import {typography} from '@utils/typography';
 
 type Props = SearchProps & {
-    defaultHeight: number;
-    largeHeight: number;
-    scrollValue?: Animated.SharedValue<number>;
+    topMargin: AnimatedStyleProp<ViewStyle>;
```
Comment (enahum) [inline app/components/navigation_header/search.tsx:14]: perhaps change the name to `topStyle` as this is no longer a value

File: app/components/post_draft/post_draft.tsx [hunk 5]
```diff
@@ -74,7 +74,7 @@ function PostDraft({
         ios: (keyboardHeight ? keyboardHeight - keyboardAdjustment : (postInputTop + insetsAdjustment)),
         default: postInputTop + insetsAdjustment,
     });
-    const autocompleteAvailableSpace = containerHeight - autocompletePosition - (isChannelScreen ? headerHeight + insets.top : 0);
+    const autocompleteAvailableSpace = containerHeight - autocompletePosition - (isChannelScreen ? headerHeight : 0);
```
Comment (enahum) [inline app/components/post_draft/post_draft.tsx:77]: why this change?

File: app/constants/view.ts [hunk 6]
```diff
@@ -16,19 +16,25 @@ export const TABLET_SIDEBAR_WIDTH = 320;
...
+export const LARGE_HEADER_TITLE_HEIGHT = 60;
 export const HEADER_WITH_SUBTITLE = 24;
 export const KEYBOARD_TRACKING_OFFSET = 72;
-export const HEADER_SEARCH_HEIGHT = SEARCH_INPUT_HEIGHT + 5;
-export const HEADER_SEARCH_BOTTOM_MARGIN = 10;
+export const UNLOCKED_SEARCH_BOTTOM_MARGIN = Platform.select({android: -14, default: -14});
+export const LOCKED_SEARCH_BOTTOM_MARGIN = Platform.select({android: 11, default: 0});
+
+export const TABLET_SEARCH_HEIGHT_COLLAPSED = Platform.select({android: 88, default: 98}); // 98 total, insets=24, 74
+export const TABLET_SEARCH_HEIGHT_EXPANDED = Platform.select({android: 145, default: 182}); // 182 total, insets=24, 158
```
Comment (enahum) [inline app/constants/view.ts:26]: can you explain this values?

File: app/hooks/header.ts [hunk 7]
```diff
@@ -17,53 +17,61 @@ type HeaderScrollContext = {
...
-        return ViewConstants.TABLET_HEADER_HEIGHT;
+        const tabletHeight = hasSearch ? ViewConstants.TABLET_SEARCH_HEIGHT_COLLAPSED : ViewConstants.TABLET_HEADER_HEIGHT;
+        return tabletHeight;
     }
 
-    if (Platform.OS === 'ios') {
-        return ViewConstants.IOS_DEFAULT_HEADER_HEIGHT;
+    let height = isIOS ? ViewConstants.IOS_DEFAULT_HEADER_HEIGHT : ViewConstants.ANDROID_DEFAULT_HEADER_HEIGHT;
+    if (hasSearch) {
+        height += isIOS ? 7 : 53;
```
Comment (enahum) [inline app/hooks/header.ts:32]: can you explain this values?

File: app/screens/channel/header/header.tsx [hunk 8]
```diff
@@ -75,13 +74,12 @@ const ChannelHeader = ({
     const theme = useTheme();
     const styles = getStyleSheet(theme);
     const defaultHeight = useDefaultHeaderHeight();
-    const insets = useSafeAreaInsets();
     const callsAvailable = callsEnabledInChannel && !callsFeatureRestricted;
 
     const isDMorGM = isTypeDMorGM(channelType);
     const contextStyle = useMemo(() => ({
-        top: defaultHeight + insets.top,
```
Comment (enahum) [inline app/screens/channel/header/header.tsx:83]: why this change?

File: app/screens/channel/channel.tsx [hunk 9]
```diff
@@ -94,7 +93,7 @@ const Channel = ({
         return () => back?.remove();
     }, [componentId, isTablet]);
 
-    const marginTop = defaultHeight + (isTablet ? insets.top : 0);
+    const marginTop = defaultHeight;
```
Comment (enahum) [inline app/screens/channel/channel.tsx:96]: why this change?

File: app/screens/global_threads/index.tsx [hunk 10]
```diff
@@ -39,13 +38,13 @@ const GlobalThreads = ({componentId}: Props) => {
     const [tab, setTab] = useState<GlobalThreadsTab>('all');
 
     const containerStyle = useMemo(() => {
-        const marginTop = defaultHeight + insets.top;
+        const marginTop = defaultHeight;
```
Comment (enahum) [inline app/screens/global_threads/index.tsx:41]: why this changes?

File: app/screens/home/recent_mentions/recent_mentions.tsx [hunk 11]
```diff
@@ -87,8 +86,7 @@ const RecentMentionsScreen = ({mentions, currentTimezone, isTimezoneEnabled}: Pr
     }, [serverUrl, isFocused]);
 
     const {scrollPaddingTop, scrollRef, scrollValue, onScroll, headerHeight} = useCollapsibleHeader<FlatList<string>>(true, onSnap);
-    const paddingTop = useMemo(() => ({paddingTop: scrollPaddingTop - insets.top, flexGrow: 1}), [scrollPaddingTop, insets.top]);
-    const scrollViewStyle = useMemo(() => ({top: insets.top}), [insets.top]);
+    const paddingTop = useMemo(() => ({paddingTop: scrollPaddingTop, flexGrow: 1}), [scrollPaddingTop]);
```
Comment (enahum) [inline app/screens/home/recent_mentions/recent_mentions.tsx:89]: why this change?

File: app/screens/home/saved_messages/saved_messages.tsx [hunk 12]
```diff
@@ -88,8 +87,7 @@ function SavedMessages({posts, currentTimezone, isTimezoneEnabled}: Props) {
     }, [serverUrl, isFocused]);
 
     const {scrollPaddingTop, scrollRef, scrollValue, onScroll, headerHeight} = useCollapsibleHeader<FlatList<string>>(true, onSnap);
-    const paddingTop = useMemo(() => ({paddingTop: scrollPaddingTop - insets.top, flexGrow: 1}), [scrollPaddingTop, insets.top]);
-    const scrollViewStyle = useMemo(() => ({top: insets.top}), [insets.top]);
+    const paddingTop = useMemo(() => ({paddingTop: scrollPaddingTop, flexGrow: 1}), [scrollPaddingTop]);
```
Comment (enahum) [inline app/screens/home/saved_messages/saved_messages.tsx:90]: same as recent mentions.. what changed here?

File: app/screens/home/search/search.tsx [hunk 13]
```diff
@@ -266,6 +271,16 @@ const SearchScreen = ({teamId}: Props) => {
...
     ), [cursorPosition, handleTextChange, searchValue, autocompleteMaxHeight, autocompletePosition]);
 
+    const onFlatLayout = useCallback(() => {
+        if (clearRef.current) {
+            scrollRef.current?.scrollToOffset({
+                offset: headerOffset,
+                animated: false,
+            });
+            clearRef.current = false;
+        }
```
Comment (enahum) [inline app/screens/home/search/search.tsx:281]: why is this needed?

**Review by enahum — COMMENTED** [review]
a few questions before approving

File: app/components/post_draft/post_draft.tsx (hunk 5 above)
Comment (jfrerich) [reply app/components/post_draft/post_draft.tsx:77]: @elias, for all of the components that have `insets.top` removed, I did a refactor.  

Each component that used `useDefaultHeader` was having to account for `insets.top` inside the component.  I added the `insets.top` to the return value from `useDefaultHeader` so that each component doesn't need to account for that value.

File: app/screens/global_threads/index.tsx (hunk 10 above)
Comment (jfrerich) [reply app/screens/global_threads/index.tsx:41]: This is from refactor in eaad1c3

File: app/screens/home/recent_mentions/recent_mentions.tsx (hunk 11 above)
Comment (jfrerich) [reply app/screens/home/recent_mentions/recent_mentions.tsx:89]: This is from refactor in eaad1c3

File: app/components/post_draft/post_draft.tsx (hunk 5 above)
Comment (jfrerich) [reply app/components/post_draft/post_draft.tsx:77]: This is from refactor in eaad1c3

File: app/screens/home/search/search.tsx (hunk 13 above)
Comment (jfrerich) [reply app/screens/home/search/search.tsx:283]: I added a comment to this code. 

```ts
    // when clearing the input from the search results, scroll the initial view
    // back to the top so the header is in the collapsed state

```

File: app/constants/view.ts (hunk 6 above)
Comment (jfrerich) [reply app/constants/view.ts:26]: These are the values for the expanded and collapsed heights.  I'll fix these and account for the component heights needed to get the values in stead of hard coding them.  I'll get those fixed and ask for a re-review.

**Comment by jfrerich** [comment]
@matthewbirtch tickets created for follow updated

Autocomplete opens when recent search selected https://mattermost.atlassian.net/browse/MM-47766
Empty results Margins https://mattermost.atlassian.net/browse/MM-47765

File: app/components/post_draft/post_draft.tsx (hunk 5 above)
Comment (enahum) [reply app/components/post_draft/post_draft.tsx:77]: lol my gh username is @enahum ;)

Comment (jfrerich) [reply app/components/post_draft/post_draft.tsx:77]: `elias username` sorry for tagging you in this PR if you see this :) 
  
  @enahum 🤣.

File: app/screens/channel/channel.tsx (hunk 9 above)
Comment (jfrerich) [reply app/screens/channel/channel.tsx]: After checking the result of this change, I've corrected this code to reflect the following change. Because defaultHeader includes the `insets.top` measurement I needed subtract in from both values in the ternary operator.  I also verified the new messags toast appears again. It was misaligned.

image

With the change above. 
image
image

File: app/screens/gallery/header/index.tsx [hunk 14]
```diff
@@ -43,8 +43,9 @@ const edges: Edge[] = ['left', 'right', 'top'];
 const AnimatedSafeAreaView = Animated.createAnimatedComponent(SafeAreaView);
 
 const Header = ({index, onClose, style, total}: Props) => {
+    const insets = useSafeAreaInsets();
     const {width} = useWindowDimensions();
-    const height = useDefaultHeaderHeight();
+    const height = useDefaultHeaderHeight() - insets.top;
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply app/screens/gallery/header/index.tsx:48]: useDefaultHeader includes insets.top with the refactor. 

by subtracting the insets.top out, we have the same result as before the refactor.
image

If I don't subtract the insets.top out we get the following:
image

File: app/hooks/header.ts (hunk 7 above)
Comment (jfrerich) [reply app/hooks/header.ts]: I made some changes to help clear this code up.  

The new `hooks/header.ts` calculations for `defaultHeader` and `largeHeaderHeight` now have `if (hasSearch)` block that will return the values for the collapsed and expanded search headers heights. 

Looking in `constants/view.ts` the calculations are made in there and a single const is used for each height.

image

`UNLOCKED_SEARCH_BOTTOM_MARGIN` and `LOCKED_SEARCH_BOTTOM_MARGIN` are the total margin calculations to get the search input located above the intial and results components (with 14x margin added per UX)

`TABLET_SEARCH_HEIGHT_COLLAPSED` and `MOBILE_SEARCH_HEIGHT_COLLAPSED` are the absolute heights for the collapsed states

`TABLET_SEARCH_MARGIN_EXPANDED` is the additional margin added to the expanded search header.
The total expanded search height is calculated in `useLargeHeaderHeight` by adding `defaultHeight + ViewConstants.LARGE_HEADER_TITLE_HEIGHT + TABLET_SEARCH_MARGIN_EXPANDED` where 
* defaultHeight - is based on the `SEARCH_INPUT_HEIGHT` + `14px margin` + `insets.top`
* `LARGE_HEADER_TITLE_HEIGHT` is the height of the large text height
* `TABLET_SEARCH_MARGIN_EXPANDED` is the additional margin request from UX

**Comment by jfrerich** [comment]
FYI, this is still not correct.  @matthewbirtch, I have a UX question.  In the expanded state, I feel like switching through the search, recent mentions, and saved messages screen should all have the same height. so there is not flickering between the screen.  I will design to that criteria.  @elias, this should help with the constants used to calculate the search expanded and collapsed heights.

**Comment by matthewbirtch** [comment]
> FYI, this is still not correct. @matthewbirtch, I have a UX question. In the expanded state, I feel like switching through the search, recent mentions, and saved messages screen should all have the same height. so there is not flickering between the screen. I will design to that criteria. @elias, this should help with the constants used to calculate the search expanded and collapsed heights.

@jfrerich I'm not 100% sure exactly what you mean, but the tabs should remain in whatever state they were left in last so that when you switch between them they retain the previous state and scroll you left it at. We should not reset the header height or reset the scroll when switching between tabs. Is this what you're proposing?

Or are you just suggesting that the search expanded top bar should be the same height as the recent mentions expanded top bar. If so, I agree! They should be consistent height in their expanded states
